File: examples/python/retry.py
"""Idempotent retry with exponential backoff and jitter.

Run:  PUSHFY_API_TOKEN=... python3 retry.py

Retry only transient failures (5xx / network via ApiError, and 429 via
RateLimitError). Reuse the SAME ext_id on every attempt so a retry can never
create a duplicate message or a double charge. Never retry AuthenticationError
or InvalidRequestError — those won't succeed on a repeat.

WARNING: a send that times out may have SUCCEEDED server-side. For strict
safety, look up the status by ext_id before resending. Because we hold ext_id
fixed here, the platform can dedupe a genuine repeat.
"""

import os
import random
import time

from pushfy import (
    Pushfy,
    ApiError,
    AuthenticationError,
    InvalidRequestError,
    RateLimitError,
)

MAX_ATTEMPTS = 5
BASE_DELAY = 1.0   # seconds
MAX_DELAY = 30.0   # cap the backoff


def backoff_delay(attempt):
    """Exponential backoff (1s, 2s, 4s, 8s...) with full jitter."""
    capped = min(MAX_DELAY, BASE_DELAY * (2 ** attempt))
    return random.uniform(0, capped)


def send_with_retry(pushfy, **kwargs):
    last_error = None
    for attempt in range(MAX_ATTEMPTS):
        try:
            return pushfy.sms.send(**kwargs)

        except (AuthenticationError, InvalidRequestError):
            # Permanent — retrying won't help. Re-raise immediately.
            raise

        except (RateLimitError, ApiError) as err:
            # Transient — back off and try again with the same ext_id.
            last_error = err
            if attempt == MAX_ATTEMPTS - 1:
                break
            delay = backoff_delay(attempt)
            print("Attempt %d failed (%s); retrying in %.2fs"
                  % (attempt + 1, type(err).__name__, delay))
            time.sleep(delay)

    raise last_error


def main():
    pushfy = Pushfy(api_token=os.environ["PUSHFY_API_TOKEN"])

    result = send_with_retry(
        pushfy,
        to="5511999999999",
        text="Hello from Pushfy",
        ext_id="welcome-001",   # stable across every retry
    )
    print("Accepted:", result)


if __name__ == "__main__":
    main()

File: examples/php/retry.php
<?php

declare(strict_types=1);

// Idempotent retry with exponential backoff + jitter.
//   PUSHFY_API_TOKEN=... php retry.php
//
// Golden rule: reuse a STABLE extId across attempts and, before resending
// after a timeout, ask the API whether the message already went out — this
// prevents double-charging. We only retry transient failures (5xx / network
// / rate limit); 4xx and auth errors are fatal and re-thrown immediately.

require __DIR__ . '/vendor/autoload.php';

use Pushfy\Pushfy;
use Pushfy\Exception\RateLimitException;
use Pushfy\Exception\ApiException;
use Pushfy\Exception\PushfyException;

$pushfy = new Pushfy([
    'apiToken' => getenv('PUSHFY_API_TOKEN') ?: '',
]);

/**
 * Send one SMS with bounded, idempotent retries.
 *
 * @param array{to: string, text: string, extId: string} $params
 * @return mixed
 * @throws PushfyException when retries are exhausted or the error is fatal.
 */
function sendWithRetry(Pushfy $pushfy, array $params, int $maxAttempts = 5)
{
    $extId = $params['extId']; // STABLE across every attempt.

    for ($attempt = 1; $attempt <= $maxAttempts; $attempt++) {
        try {
            // Before a resend, check if a previous attempt already succeeded.
            if ($attempt > 1) {
                $status = $pushfy->messages->status(['extId' => $extId]);
                if (!empty($status)) {
                    fwrite(STDERR, "extId {$extId} already accepted — not resending.\n");
                    return $status;
                }
            }

            return $pushfy->sms->send($params);
        } catch (RateLimitException $e) {
            $delay = $e->getRetryAfter() ?? backoffSeconds($attempt);
            rethrowIfLast($e, $attempt, $maxAttempts);
            fwrite(STDERR, "Rate limited, waiting {$delay}s (attempt {$attempt}).\n");
            sleep((int) $delay);
        } catch (ApiException $e) {
            // 5xx / network / timeout — transient.
            $delay = backoffSeconds($attempt);
            rethrowIfLast($e, $attempt, $maxAttempts);
            fwrite(STDERR, "Transient error (status {$e->getStatus()}), waiting {$delay}s (attempt {$attempt}).\n");
            sleep((int) $delay);
        }
        // Any other PushfyException (auth/4xx) is fatal — let it propagate.
    }

    throw new ApiException('Retries exhausted for extId ' . $extId, 0);
}

/** Exponential backoff with full jitter: base 0.5s, capped at 30s. */
function backoffSeconds(int $attempt): float
{
    $cap = 30.0;
    $base = 0.5 * (2 ** ($attempt - 1));
    $window = min($cap, $base);
    return mt_rand(0, (int) ($window * 1000)) / 1000; // full jitter
}

function rethrowIfLast(PushfyException $e, int $attempt, int $maxAttempts): void
{
    if ($attempt >= $maxAttempts) {
        throw $e;
    }
}

try {
    $result = sendWithRetry($pushfy, [
        'to'    => '5511999999999',
        'text'  => 'Reliable delivery via idempotent retry',
        'extId' => 'retry-demo-fixed-key', // reuse the same key on rerun
    ]);
    echo "Delivered/accepted:\n";
    echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE), "\n";
} catch (PushfyException $e) {
    fwrite(STDERR, sprintf(
        "Gave up: status=%d code=%s\n",
        $e->getStatus(),
        (string) $e->getErrorCode()
    ));
    exit(1);
}
